### PR TITLE
fix(scan): let a complete rescan clear the ids it rematched

### LINKED ISSUE #4346: [Bug] Complete rescan does not wipe external metadata IDs when a selected provider finds no match
**Describe the bug**

RomM's scan reference tells users that a Complete rescan resets external IDs:

> **Complete Rescan:** Rescans and rematches all games in the selected platforms (slowest).
> This will wipe all existing metadata matches, including the external IDs, and attempt to
> match them again, like on a fresh scan.

(`frontend/src/locales/en_US/scan.json`, key `scan.scan-types-info`; repeated in
`frontend/src/v2/components/Scan/ScanInfoDialog.vue`.)

That only holds when the provider finds a new match. If a selected provider searches and comes back
empty, the old (wrong) ID and its metadata blob are silently kept. So the one scan type documented as
the way to undo a bad match is exactly the one that can't undo it, and this bites hardest on obscure,
homebrew or oddly-named ROMs, which are where bad matches happen in the first place.

**To Reproduce**

1. Pick a ROM whose filename no provider can match (or rename one so it can't be matched).
2. Set a deliberately wrong IGDB ID on it: Edit ROM → Metadata IDs → set `igdb_id`, save.
3. Run a Complete rescan over that platform with IGDB **selected**.
4. Open the ROM again.

**Expected behavior**

`igdb_id` and `igdb_metadata` are cleared and the ROM shows as unmatched for IGDB, "like on a fresh scan".

**Actual behavior**

`igdb_id` and `igdb_metadata` are unchanged. The wrong ID survives the rescan.

**Root cause**

`backend/handler/scan_handler.py` clears IDs on a COMPLETE rescan only for sources the user *deselected*:

```python
if not newly_added and scan_type == ScanType.COMPLETE:
    for source, fields in metadata_handlers.items():
        if source not in metadata_sources:
            rom_attrs[fields["id_field"]] = None
```

Selected sources are assumed to reset implicitly because the carry-forward block
(`if not newly_added and scan_type != ScanType.COMPLETE:`) is skipped on COMPLETE, leaving the IDs out
of `rom_attrs`. That assumption doesn't hold: `add_rom` is `session.merge`, and SQLAlchemy's merge
**leaves an attribute untouched when it is absent from the source object's `__dict__`** — it does not
write NULL. `Rom` is a plain `DeclarativeBase` (not dataclass-mapped), so unset kwargs never enter
`__dict__` at all.

Confirmed against the real model (SQLAlchemy's `_sa_*` internals filtered out):

```
attrs set on transient Rom: ['_is_identifying','fs_extension','fs_name','fs_name_no_ext','fs_name_no_tags','id','platform_id','ss_id']
igdb_id in __dict__? False | ss_id in __dict__? True
```

and through merge:

```
transient keys: ['fs_name', 'id', 'ss_id']
after merge -> igdb_id: 1289 | ss_id: None
```

The later priority-merge loop only assigns truthy values, and a source that found nothing is excluded
from `available_sources`, so no code path ever writes the ID back to NULL.

**Suggested fix**

Clear the id/blob for sources the scan actually *attempted*, in addition to deselected ones.

Note the trap: a provider can fail to run because the platform is unsupported (`platform.igdb_id` is
None, or `platform.slug` not in the provider's platform list) rather than because it searched and found
nothing. Blanket-nulling every source would wipe hand-set IDs on unmapped platforms. Tracking the
sources that entered their fetch branch in a set (safe under `asyncio.gather`) distinguishes the two.

Every provider's gate already contains `or scan_type == ScanType.COMPLETE`, so on a complete rescan
"attempted" reduces to "selected, and supported for this platform" — the set only has to record what the
existing gates already decide, rather than restating any platform predicates.

**Additional context**

Workaround today: run a Complete rescan with the offending provider **deselected** (which does null the
ID), then a second scan with it selected.

### PR DESCRIPTION
**Pull request by Spinnich**
**Description**

Fixes #4346.

A Complete rescan is documented as wiping every match, including the external IDs, and matching again "like on a fresh scan". It only did that for sources the user **deselected**. A selected provider that searched and came back empty silently kept its old (wrong) ID and metadata blob, so the one scan type meant to undo a bad match couldn't undo it. This bites hardest on homebrew and oddly-named ROMs, which is where bad matches come from in the first place.

The reset was assumed to happen implicitly: on COMPLETE the carry-forward block is skipped, so the IDs stay out of `rom_attrs`. But `add_rom` is `session.merge`, and merge leaves an attribute untouched when it's absent from the source object's `__dict__` rather than writing NULL. `Rom` is a plain `DeclarativeBase`, so unset kwargs never enter `__dict__`. Nothing downstream writes the ID back either: the priority-merge loop only assigns truthy values, and a source that found nothing isn't in `available_sources`.

The fix tracks the sources whose lookup actually ran and clears those alongside the deselected ones:

- Each provider's fetch branch records itself in an `attempted_sources` set. Every gate already contains `or scan_type == ScanType.COMPLETE`, so on a complete rescan "attempted" reduces to "selected, and supported for this platform" — no platform predicates are restated.
- A source skipped because the platform carries no ID for it (`platform.igdb_id is None`, slug not in the provider's list) never enters the set, so hand-set IDs on unmapped platforms survive. Blanket-nulling would have wiped them.
- A provider that raised is discarded from the set in the `asyncio.gather` result loop, and a ScreenScraper rate-limit refusal discards itself. Neither ruled anything out, so both keep the existing ID.
- SteamGridDB resolves after the reset block (it needs the final name), so it clears its own stale ID when its lookup came back empty.

While in that loop, the error log now names the source rather than the fallback class, since the source is right there.

Only COMPLETE is affected. Quick, Update and Unmatched scans carry IDs forward exactly as before.

**A failed lookup is not a miss**

That reset only holds up if a provider's empty result means "no match". Six report a failure as one instead, so a rescan run while any of them was down would have wiped the ID it would otherwise have confirmed, hand-set ones included. Each now reports the failure and leaves the tolerance to the caller, which is what the gather loop's discard already assumed:

- Pouët, CSDb and Demozoo stop swallowing the 503 their own `_request` raises, and Pouët's title search raises it rather than answering "no unique title".
- SteamGridDB and Flashpoint stop catching `Exception` around a whole lookup; HLTB stops catching it around its search.
- ScreenScraper answers empty for every remaining ROM once its daily-quota or rejected-credentials breaker trips, so `scan_rom` reads that state off the service and drops SS from the sources the rescan may clear. This was the sharpest of the set: an exhausted SS quota would have cleared `ss_id` across the rest of the library.

SteamGridDB resolves outside the gather, so it needs the discard at its call site. `/search/rom` keeps degrading to no thumbnail, since its cover art sits on top of matches the other providers already produced.

Unchanged and worth naming: `SteamGridDBService._request` turns a non-401 `ClientResponseError` into `{}`, so SteamGridDB answering 5xx still reads as "no results" a layer below this. Closing that changes what `/search/cover` returns on an outage, so it stays out of a scan fix.

**Testing**

New `backend/tests/handler/test_scan_complete_rescan_ids.py` (9 tests) drives the real `scan_rom` and persists through `add_rom`, so it exercises the merge behavior that caused the bug rather than mocking around it: a searched source that misses drops its stale ID, one that matches stores the new ID, an unsupported platform and a failing provider both keep theirs, deselected sources still clear, SteamGridDB clears after a miss, and the three non-COMPLETE scan types are parametrized to confirm they're untouched.

Four more cases cover the flip side: an unreachable Pouët and an unreachable SteamGridDB both keep their IDs, a tripped ScreenScraper breaker keeps `ss_id` while a selected ScreenScraper that genuinely misses still clears it. Each changed handler also gets a test that its failure propagates rather than surfacing as an empty match.

- `uv run pytest tests/handler tests/endpoints tests/adapters` — 2902 passed
- `trunk fmt && trunk check` — clean

**AI assistance disclosure**

This PR was written with AI assistance (Claude Code). The root cause analysis, implementation and tests were AI-generated and reviewed by me; the SQLAlchemy merge behavior described above was verified empirically against the real model before the fix was written.

**Checklist**

- [x] I've tested the changes locally
- [x] I've updated relevant comments
- [x] I've assigned reviewers for this PR
- [x] I've added unit tests that cover the changes

### CODE REVIEW

File: backend/handler/scan_handler.py [hunk 1]
```diff
@@ -1291,11 +1327,11 @@ async def fetch_hasheous_rom(hasheous_rom: HasheousRom) -> HasheousRom:
...
 
-    # For COMPLETE rescans, explicitly clear metadata IDs and metadata for unselected sources
-    # This ensures that when a source is no longer selected, its data is removed from the ROM
-    if not newly_added and scan_type == ScanType.COMPLETE:
+    # Reset the id and blob of every source this rescan could rematch: the ones it
+    # consulted, whose miss has to undo the old match, and the deselected ones.
+    if is_complete_rescan:
         for source, fields in metadata_handlers.items():
-            if source not in metadata_sources:
+            if source in attempted_sources or source not in metadata_sources:
```
[the remark this answers is not in the document]
Comment (Spinnich) [reply backend/handler/scan_handler.py:1347]: Valid, fixed in 24bc4c221.

Confirmed the mechanism: `_request` raises a 503, but `get_rom_by_id` caught it and answered `PouetRom(pouet_id=None)`, and `resolve_exact_title` swallowed `ConnectError`/`ReadTimeout` into "no unique title". CSDb was the same shape. So an outage was indistinguishable from a miss, the reset cleared the id, and `scene_locked` (which reads `rom_attrs` after the reset) then dropped with it.

Rather than special-case the scene sources, all three now stop swallowing, so the existing `attempted_sources.discard(source)` in the gather loop does its job. Demozoo had the same swallow in `get_rom_by_id`, so it went too. This also makes the manual-match endpoint report "Can't connect to Pouet API" instead of silently not matching, which is how IGDB already behaves there.

Auditing the rest of the loop turned up three more of the same defect, fixed in the same commit: Flashpoint (`search_games`, `get_rom_by_id`) and HLTB (`search_games`) catch `Exception` around a whole lookup, and ScreenScraper answers empty for every remaining ROM once its daily-quota or rejected-credentials breaker trips. That last one was the worst of the set: an exhausted SS quota would have wiped `ss_id` across the rest of the library.

File: backend/handler/scan_handler.py [hunk 2]
```diff
@@ -1500,6 +1537,10 @@ async def fetch_sgdb_details(playmatch_rom: PlaymatchRomMatch) -> SGDBRom:
             )
             if ranked[0] == MetadataSource.SGDB:
                 rom_attrs["url_cover"] = sgdb_cover
+    elif is_complete_rescan and MetadataSource.SGDB in attempted_sources:
+        # SteamGridDB resolves after the reset block above, so it has to clear
+        # its own stale id when its lookup came back empty.
+        rom_attrs["sgdb_id"] = None
```
[the remark this answers is not in the document]
Comment (Spinnich) [reply backend/handler/scan_handler.py]: Valid, fixed in 24bc4c221.

`get_details_by_names` and `get_rom_by_id` both caught `Exception` around the whole lookup and returned `SGDBRom(sgdb_id=None)` (the docstring promised never to raise), so a network error, timeout or rejected key was indistinguishable from a name SteamGridDB has no art for.

Both now report the failure, and each caller decides its own tolerance. SteamGridDB resolves outside the `asyncio.gather`, so `scan_rom` catches it at the call site and discards SGDB from `attempted_sources`, matching what the gather loop does for every other provider. `/search/rom` keeps degrading to no thumbnail, since the cover art there sits on top of matches the other providers already produced, and failing the search over it would be a worse trade.

One boundary worth naming: `SteamGridDBService._request` turns a non-401 `ClientResponseError` into `{}`, so SteamGridDB answering 5xx still reads as "no results" a layer below this. Closing that changes what `/search/cover` returns on an outage, so I left it out of this PR rather than widen an API contract inside a scan fix. The realistic outages (host unreachable, timeout, bad credentials) are covered.

**Review by gantoine — APPROVED** [review]
